crates/runtime/src/externref.rs

```diff
@@ -393,6 +461,704 @@ impl Deref for VMExternRef {
     }
 }
 
+type TableElem = UnsafeCell<Option<VMExternRef>>;
+
+/// A table that over-approximizes the set of `VMExternRef`s that any Wasm
+/// activation on this thread is currently using.
+///
+/// Under the covers, this is a simple bump allocator that allows duplicate
+/// entries. Deduplication happens at GC time.
+#[repr(C)]
+pub struct VMExternRefActivationsTable {
+    /// Bump-allocation finger within the current chunk.
+    ///
+    /// NB: this is an `UnsafeCell` because it is read from and written to by
+    /// compiled Wasm code.
+    next: UnsafeCell<NonNull<TableElem>>,
```

Comment:
How is this field used from wasm code? I tried to read over but couldn't find much... It also looked like `VMOffsets` contained information about the `end` field below too, but it's not an `UnsafeCell`?

Reply:
Ah ok I confirmed on Zulip, but it sounds like this is planned to get used in the future. I presume that the idea is that with `table.get` you'd have a fast path for using these two fields to inline the `try_insert` method, and you'd fallback to an intrinsic for out-of-line growth and such?

Reading over all this dealing with chunks and such feels a bit overly complicated. Could this perhaps be simplified a bit? I'm thinking that this could perhaps be restructured to something that looks like:

```rust
 pub struct VMExternRefActivationsTable {
    // finter into `fast_path_storage`
    next: UnsafeCell<NonNull<TableElem>>,
    // never changes after creation
    end: NonNull<TableElem>,
    // used by `try_insert` and stored via `next` in jit code
    fast_path_storage: Box<[TableElem]>,
    
    imprecise_roots: HashSet<VMExternRef>,
    precise_roots: HashSet<VMExternRef>,
}
```

where here `fast_path_storage` is allocated once and never changes. When an overflow of that table happens it's drained and everything is moved into `imprecise_roots`. That way `imprecise_roots` is a growing set which deduplicates things (in case you `table.get` the same thing a bunch of times) and also handles reallocation for us (we don't need to manually double capacities and such).

On a GC we'd fill in `precise_roots` and then we'd `mem::swap` the precise/imprecise sets and then clear the imprecise one. Maybe with some other trickery around optimizing reference counts or something like that.

I'm mostly hoping that we can simplifiy the `chunks` list because:

* It seems somewhat complicated to manage, especially if we're trying to be clever about capacities and double them
* I'm a bit worried about the segmented-stacks thrashing problem where you keep allocating a doubly-bigger chunk but then freeing it, when it'd be more efficient to just double the size of the internal chunk.
* Roots aren't deduplicated in the chunks so if you pass around the same `VMExternRef` a bunch it may accidentally fill them up quite a lot.

With a persistent hash set I think it'd solve the duplication/capacity issues? Anyway curious what you think about this!

Reply:
This is a good idea, however

* we may still want to have the fast-path table grow shrink based on the precise size of the last root set (probably fine to skip for now, but we'll likely want to investigate this in the future)

* we lose the ability to reclaim memory once we are no longer using excessive capacity in these sets (again, probably fine for now, but we'll likely want to handle this sometime in the future)

Reply:
Switching `fast_path_storage` to a `Vec<T>` would be an easy way to make the fast-path-table growable (that's a good point, and one thing I was curious how often it would come into play). For shrinking we could periodically call `HashSet::shrink_to_fit` I think?

Reply:
Yeah. Would be nice if `HashSet::shrink_to` was stable, so we could shrink capacity by half when the length is a quarter of the capacity, and keep the amortized O(1). We could do `shrink_to_fit` followed by `reserve` but this is suboptimal.

Anyways, I'm just gonna do the simplest thing here for now, and leave that stuff to when we actually see it in profiles.